src/common/fields-derive/src/lib.rs

```diff
@@ -0,0 +1,76 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use proc_macro2::TokenStream;
+use quote::quote;
+use syn::Data;
+use syn::DeriveInput;
+use syn::Field;
+use syn::Result;
+
+#[proc_macro_derive(Fields)]
+pub fn fields(tokens: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let input: DeriveInput = syn::parse_macro_input! {tokens};
+
+    match gen(input) {
+        Ok(tokens) => tokens.into(),
+        Err(err) => err.to_compile_error().into(),
+    }
+}
+
+fn gen(input: DeriveInput) -> Result<TokenStream> {
+    let DeriveInput {
+        attrs: _attrs,
+        vis: _vis,
+        ident,
+        generics,
+        data,
+    } = input;
+    if !generics.params.is_empty() {
+        return Err(syn::Error::new_spanned(
+            generics,
+            "generics are not supported",
+        ));
+    }
+
+    let Data::Struct(r#struct) = data else {
+        return Err(syn::Error::new_spanned(ident, "only structs are supported"));
+    };
+
+    let fields_rs = r#struct.fields;
+    let fields_rw: Vec<TokenStream> = fields_rs
+        .into_iter()
+        .map(|field_rs| {
+            let Field {
+                // We can support #[field(ignore)] or other useful attributes here.
+                attrs: _attrs,
+                ident: name,
+                ty,
+                ..
+            } = field_rs;
+            let name = name.map_or("".to_string(), |name| name.to_string());
```

Comment:
Why a struct field may have no name? 🤔

Reply:
tuple

Reply:
In PostgreSQL the names would implicitly become `f1`, `f2`, ...
```
test=# select row(9, 18);
  row   
--------
 (9,18)
(1 row)

test=# select (row(9, 18)).f1;
 f1 
----
  9
(1 row)

test=# select to_jsonb(row(9, 18));
      to_jsonb       
---------------------
 {"f1": 9, "f2": 18}
(1 row)
```

But in RisingWave this has not been supported consistently yet. Some places are empty while some places have these implicit names. Have not thought about the impact on backward compatibility if it was fixed to be pg-compatible.

No opinion on what the behavior of this place should be.